### ISSUE TITLE
Mixin attribute values should propagate through the inheritance tree just as regularly inherited values do.

### ISSUE THREAD

**Issue by finger563** [webgme/webgme#1432]

**Comment by pmeijer** [comment]
The better approach here would be to introduce a default value for attributes as part of the meta definition of the attribute. Propagating values defined in the mixin nodes (outside of the meta) would be against the mixin concept.

**Comment by finger563** [comment]
Wouldn't those default values then propagate still? How is that different from the mixin values propagating? (since the mixin default value would propagate)

**Comment by kecso** [comment]
Mixin values should not propagate. That is why we have this concept instead of a 'pure' multiple inheritance. No matter how you propagate data, if you propagate from multiple sources, you need to order them (or handle the possible collision somehow). That rule was incompatible with and easy and lean visual modelling paradigm.

**Comment by pmeijer** [comment]
@finger563 The default values would propagate as part of the definition (meta-rule) of the attribute.

**Comment by finger563** [comment]
@pmeijer wouldn't that still have an issue if two mixins that were used by the same node had same attribute names and be the issue that @kecso is talking about? Personally I agree it gets hairy, but providing an interface to manage complexity *only when it arises* seems like a good solution, and falls in line with what webGME does in the general case.  I.e. I don't have to worry about multiple inheritance issues or deal with clunky UI that is designed to handle these things _unless I run into a case where it cannot be avoided_.

In the end of the day it is more headache and overhead, so simply avoiding it by not dealing with mixin attribute value inheritance is fine.  Along those lines, how does the meta-evaluator deal with mixin attribute inheritance?  Since no default value is provided will newly created nodes automatically fail the meta-evaluation?  Esp. if bounds or regex is applied to the attribute?

**Comment by pmeijer** [comment]
I realize that default values wouldn't add anything. If `B` mixes in `A`, which defines attribute `a`. Then setting the value of `a` at `B` would act as the default value.
About the evaluator:
Currently if there is no value defined - it doesn't complain.

**Comment by pmeijer** [comment]
@finger563 
```
Personally I agree it gets hairy, but providing an interface to manage complexity only when it arises seems like a good solution, and falls in line with what webGME does in the general case. I.e. I don't have to worry about multiple inheritance issues or deal with clunky UI that is designed to handle these things unless I run into a case where it cannot be avoided.

In the end of the day it is more headache and overhead, so simply avoiding it by not dealing with mixin attribute value inheritance is fine.
```
If I understand you correctly, then dealing with this is already possible.. Consider this example:

![mixinattr1](https://user-images.githubusercontent.com/6518904/29622727-d38723a4-87e9-11e7-8b66-eb8b79502e61.PNG)

At this point there is a colliding definition of the `dup` attribute for `C`. Since `48596efe-17...` guid is "less" than `86a184ff-29...`'s the definition for `dup` at `C` is resolved to `48596efe-17...`'s boolean definition. Since this is uncontrollable and potentially gives unexpected results it's reported as a meta-inconsistency warning.

![mixinattr2](https://user-images.githubusercontent.com/6518904/29622886-39548690-87ea-11e7-9557-0487ff48e09d.PNG)

By explicitly adding a definition for float at `C` the `dup` becomes a float and there are no longer any complaints about inconsistencies.

I'm closing the issue - but feel free to open again if you don't agree.
